nixd/lib/Eval/AttrSetProvider.cpp

```diff
@@ -160,6 +160,61 @@ void fillOptionDescription(nix::EvalState &State, nix::Value &V,
   }
 }
 
+std::vector<std::string> completeNames(nix::Value &Scope,
+                                       const nix::EvalState &State,
+                                       std::string_view Prefix) {
+  int Num = 0;
+  std::vector<std::string> Names;
+
+  // FIXME: we may want to use "Trie" to speedup the string searching.
+  // However as my (roughtly) profiling the critical in this loop is
+  // evaluating package details.
+  // "Trie"s may not beneficial because it cannot speedup eval.
+  for (const auto *AttrPtr : Scope.attrs()->lexicographicOrder(State.symbols)) {
+    const nix::Attr &Attr = *AttrPtr;
+    const std::string_view Name = State.symbols[Attr.name];
+    if (Name.starts_with(Prefix)) {
+      ++Num;
+      Names.emplace_back(Name);
+      // We set this a very limited number as to speedup
+      if (Num > MaxItems)
+        break;
+    }
+  }
+  return Names;
+}
+
+std::optional<ValueDescription> describeValue(nix::EvalState &State,
+                                              nix::Value &V) {
+  if (V.isPrimOp()) {
+    const auto *PrimOp = V.primOp();
+    assert(PrimOp);
+    return ValueDescription{
+        .Doc = PrimOp->doc ? std::string(PrimOp->doc) : "",
+        .Arity = static_cast<int>(PrimOp->arity),
+        .Args = PrimOp->args,
+    };
+  } else if (V.isLambda()) {
+    auto *Lambda = V.payload.lambda.fun;
+    assert(Lambda);
+    return ValueDescription{
+        .Doc =
+            [&]() {
+              const auto DocComment = Lambda->docComment;
+              if (DocComment) {
+                return DocComment.getInnerText(State.positions);
+              }
+              return std::string();
```

Comment:
In this case, maybe just return `std::nullopt` also, because value description does not have any useful information.